packages/ui/Toggleable/useToggle.js

```diff
@@ -0,0 +1,77 @@
+import React from 'react';
+
+import useControlledState from '../useControlledState';
+
+import Toggle from './Toggle';
+
+/**
+ * Creates one instance of the class, `Ctor` with the provided `args`, for the life of the
+ * component.
+ *
+ * @param {Function} Ctor Class constructor
+ * @param  {...any} args Arguments to pass to the constructor
+ * @private
+ */
+function useClass (Ctor, ...args) {
+	const ref = React.useRef(null);
+	ref.current = ref.current || new Ctor(...args);
+
+	return ref.current;
+}
+
+/**
+ * Configuration for `useToggle`
+ *
+ * @typedef {Object} useToggleConfig
+ * @memberof ui/Toggleable
+ * @property {Boolean}  [defaultSelected = false] Initial state of the toggle
+ * @property {Boolean}  [disabled = false]        Disables updating the state
+ * @property {Boolean}  [prop = "selected"]       The key used to pass the current value back
+ *                                                through the `onToggle` callback.
+ * @property {Boolean}  [selected = false]        Current state of the toggle
+ * @property {Function} [onToggle]                Called when the state is changed
+ * @private
+ */
+
+/**
+ * Configuration for `useToggle`
+ *
+ * @typedef {Object} useToggleInterface
+ * @memberof ui/Toggleable
+ * @property {Boolean}  selected   Current state of the toggle
+ * @property {Function} activate   Sets the current state to `true`
+ * @property {Function} deactivate Sets the current state to `false`
+ * @property {Function} toggle     Toggles the current state to the opposite value
+ * @private
+ */
+
+/**
+ * Manages a boolean state value.
+ *
+ * The value may either be toggled via the `onToggle` or explicitly set via `onActivate` and `onDeactivate`. The initial value can be set using the `defaultSelected` option
+ * @param {useToggleConfig} config Configuration options
+ * @returns {useToggleInterface}
+ * @private
+ */
+function useToggle (config = {}) {
+	const toggle = useClass(Toggle, config);
+	const state = useControlledState(
+		config.defaultSelected,
+		config.selected,
+		typeof config.selected !== 'undefined'
+	);
+
+	toggle.setContext(...state);
+
+	return {
+		activate: toggle.handleActivate,
+		deactivate: toggle.handleDeactivate,
+		toggle: toggle.handleToggle,
+		selected: toggle.value
+	};
+}
```

Comment:
Hello @ryanjduffy 

Have you ever considered using useEvent in 'react-dom'?

The useEvent is in the master branch.
https://github.com/facebook/react/blob/master/packages/react-dom/src/client/ReactDOM.js#L217
But it is not included in any release yet. So it could be too early to apply it to our libraries.

I haven't executed any code with the useEvent. But according to the useEvent API, it seems to me that we don't have to return any event handlers in a custom hook with the useEvent to pass them to a DOM element if we could receive the ref for the DOM element.
So the `useToggle` could be

```
function useToggle (ref, config = {}) {
	const toggle = useClass(Toggle, ref, config);
	const state = useControlledState(
		config.defaultSelected,
		config.selected,
		typeof config.selected !== 'undefined'
	);

	toggle.setContext(...state);

	// no return values
}
```

Then the `toggle` could be
```
class Toggle {
	constructor (ref, config) {
		const {activate, deactivate, eventProps, prop, toggle, toggleProp} = config;

		// remapping to props for better compatibility with core/handle and binding
		this.props = config;
		this.context = {};

		if (config.activate) {
			const forwardActivate = forwardWithEventProps(config.activate);
			const activate = ReactDOM.unstable_useEvent(activate);

			React.useEffect(() => {
				activate.setListener(ref.current, (ev) => {
					if (this.activate()) forwardActivate(ev, this.props);
				});
			});
		}
		...
	}
}
```

Please fix me If I understand incorrectly.

Reply:
I'm not familiar with that API so I can't say for sure. It seems to only work for native events (due to the need for a ref) which would be challenging for our use cases (e.g. toggling on `onOpen`/`onClose`). It could be useful for managing internal event listeners for a component though so we'll have to keep an eye on it. Thanks for sharing!

Reply:
I agree with Ryan.  We don't know enough about consumers and aren't consistent enough with refs that we could be sure the events end up on the right element.  And, as Ryan says, many of the current use cases involve non-native events.

Reply:
Yes. It looks this API works with native events. Thank you for your opinions. Let's keep an eye on it.